vayesta/core/bosonic_bath/projected_interactions.py

```diff
@@ -0,0 +1,238 @@
+                ca = cb = c
+            else:
+                ca, cb = c
+
+            # Want -C_{nkc}<pk|cq> = - C_{nkc} V_{Lpc} V_{Lkq}
+
+            def _gen_exchange_spinchannel_contrib(l, c, rbos):
+                la_loc = np.tensordot(l, c, axes=(2, 0)) # "Lab,bp->Lap"
+                temp = einsum("Lkp,Lcq->kcpq", la_loc, la_loc)
+                contrib = einsum("kcpq,nkc->npq", temp, rbos)
+                return contrib
+
+            for i,(blk, lab) in enumerate(self._loop_df()):
+                if blk is not None:
+                    couplings_exchange[0] = couplings_exchange[0] + _gen_exchange_spinchannel_contrib(lab, ca, rbos_a)
+                    couplings_exchange[1] = couplings_exchange[1] + _gen_exchange_spinchannel_contrib(lab, cb, rbos_b)
+                else:
+                    raise NotImplementedError("CDERI_neg contributions to bosons not yet supported")
+
+        couplings = [x+y+z for x, y, z in zip(couplings_coulomb, couplings_exchange, couplings_fock)]
+        return couplings
+
+    def gen_nonconserving(self, couplings):
+        """Generate the particle number non-conserving part of the bosonic Hamiltonian."""
+        rbos_a, rbos_b = self.cluster.bosons.coeff_3d_ao
+        pocc = self.get_cluster_projectors()[0]
+        # This is the normal-ordered contribution, arising from non-canonical HF references.
+        contrib = einsum("npq,pq->n", rbos_a, self.fock_a) + einsum("npq,pq->n", rbos_b, self.fock_b)
+        # This arises from the transformation of the occupied term out of normal ordering
+        contrib -= einsum("npq,pq->n", couplings[0], pocc[0]) + einsum("npp,pp->n", couplings[1], pocc[1])
+        return contrib
+
+    def get_cluster_projectors(self):
+        """Get the projectors in the cluster basis into the occupied and virtual subspaces."""
+
+        def _get_cluster_spinchannel(c, co, cv):
+            so = dot(c.T, self.overlap, co)
+            po = dot(so, so.T)
+            sv = dot(c.T, self.overlap, cv)
+            pv = dot(sv, sv.T)
+            return po, pv
+
+        if self.cluster.c_active_occ[0].ndim == 1:
+            poa, pva = _get_cluster_spinchannel(self.cluster.c_active, self.cluster.c_active_occ, self.cluster.c_active_vir)
+            pob, pvb = poa, pva
+        else:
+            poa, pva = _get_cluster_spinchannel(self.cluster.c_active[0], self.cluster.c_active_occ[0],
+                                                self.cluster.c_active_vir[0])
+            pob, pvb = _get_cluster_spinchannel(self.cluster.c_active[1], self.cluster.c_active_occ[1],
+                                                self.cluster.c_active_vir[1])
+        return (poa, pob), (pva, pvb)
+
+    def _loop_df(self, blksize=None):
+        nao = self.mf.mol.nao
+        df = self.mf.with_df
+        naux = self.naux
+        if blksize is None:
+            blksize = int(1e9 / naux * nao * nao * 8)
+        # PBC:
+        if hasattr(df, 'sr_loop'):
```

Comment:
Is this always just grabbing the integrals at the gamma point? `pbc.df.DF` has `loop` implemented to do this, so that gamma point DF objects are compatible with molecular post-HF methods

Reply:
Yeah, I think this is just grabbing the gamma point integrals; I just extracted it from `get_cderis` in `core/eris.py`. I'm assuming we could similarly refactor that function then?

Reply:
Looking at that code the difference is that this also treats non positive definite eris (ie. 2D pbc systems); while it's very similar functionality I think we need to be able to treat that case, so sadly can't make that simplification for now.